### PR TITLE
New Dataset superclass

### LINKED ISSUE #55: Consolidate branch length methods to new Dataset subclass
Many Dataset subclasses have duplicated methods like `export_branch_lengths`, and the branch lengths setter. These should be moved to something called BranchLengthDataset, which can be subclassed by all the classes that currently independently implement these methods.

_Originally posted by @willdumm in https://github.com/matsengrp/netam/pull/50#discussion_r1755328935_

### PR DESCRIPTION
**Pull request by willdumm**
Addressing #55, I consolidating some branch lengths-related code into a new superclass `BranchLengthDataset`. I can't seem to find any other duplicated code.

The only opportunity for more code reuse would be to abstract methods like `to` that map over all attributes of the class and apply some function. This could be done in `BranchLengthDataset` using the `vars` builtin, but I think it might be more trouble than it's worth.

Will close #55

### CODE REVIEW

**Review by matsen — APPROVED** [review]
👍

**Review by matsen — APPROVED** [review]

File: netam/models.py [hunk 1]
```diff
@@ -59,6 +60,18 @@ def unfreeze(self):
         for param in self.parameters():
             param.requires_grad = True
 
+    def selection_factors_of_sequences(self, sequences, encoder=None):
+        if encoder is None:
+            raise ValueError("An encoder must be provided.")
+        device = next(self.parameters()).device
```
Comment (matsen) [inline netam/models.py:66]: Let's make this a `@property` .

**Review by matsen — CHANGES_REQUESTED** [review]
I appreciate what you're doing here, but there's something that doesn't quite make sense.

The `ModelBase` covers all sorts of models, e.g. SHM models, so it doesn't make sense for it to have a method `selection_factors_of_sequences`. We could call it something generic like `evaluate_sequences`, as long as all the components will make sense for SHM models and DXSM models.

**Comment by willdumm** [comment]
Changed `selection_factors_of_sequences` to `evaluate_sequences` and added device property to `ModelBase` and `AbstractBinarySelectionModel`.

**Review by matsen — APPROVED** [review]

**Pull request merged**